ports/stm/common-hal/microcontroller/Pin.c

```diff
@@ -134,3 +134,15 @@ GPIO_TypeDef * pin_port(uint8_t pin_port) {
 uint16_t pin_mask(uint8_t pin_number) {
     return 1<<pin_number;
 }
+
```

Comment:
I'd like to understand better why you needed these additions, given that I don't see them added anywhere in the rest of the PR. In general, the ST HAL does not recognize a global "pin number" system, and the GPIO banks are dangerous to iterate through due to overlap with system features (BOOT, OSC), which is why I opted for a bank vs mask approach across the STM32 port. If you have a particular implementation or algorithm that you need these for, let me know so I can take a look and potentially suggest alternatives.

Reply:
`common_hal_mcu_pin_number`, `common_hal_mcu_pin_claim` and `common_hal_mcu_pin_reset_number` are functions added already to the nrf and samd ports, and required by rgbmatrix.  See shared-{bindings,module}/rgbmatrix/RGBMatrix.c for uses.

protomatter has a fundamental assumption that there is a numbering scheme for all pins, and that given a _valid_ pin number you can get pointers to the GPIO registers (set and reset at a minimum) and the bit offset within those registers for the set or reset action.  There can of course be gaps in the numbering, and 0xff can never be a valid pin number.

Protomatter and the glue layer inside circuitpython don't iterate through GPIO banks, it just packs your two :4 bitfields int one uint8_t and then separates them out again.  It uses `pin_port()` to get the base address of that GPIO port so that it can find e.g., the right BSRR register:
```
    volatile uint16_t *_PM_portSetRegister(uint32_t pin)  {
        return (uint16_t*)&pin_port(pin / 16)->BSRR;
    }
```

If you still have concerns about the approach here, I think we'll need to talk together with Phil B. to figure out something that will come closest to satisfying everyone.

Reply:
That's alright, I get the incentive to use the bitpacking to condense things. This is one of those areas where using the HAL impacts the structural approach. It might ultimately be more efficient this way - I've considered revisiting DigitalIO and if I do I'll take a closer look at how it's handled here.